File: terrabuttler/inception.py
from colorama import Fore
from os import path
import subprocess


def inception_init_check():
    site_dir = path.realpath("site_inception")

    if (path.exists(f"{site_dir}/.terraform") and
            path.exists(f"{site_dir}/.terraform/environment")):
        return True
    return False


def inception_init_needed():
    if not inception_init_check():
        print(Fore.RED + "The initialization hasn't been made yet.\n\n"
              "Please execute the following command to initialize it:\n"
              "./run.py init")
        exit(1)


def inception_init():
    from terrabuttler.env import reload_direnv
    from terrabuttler.settings import get_settings
    site_dir = path.realpath(get_settings()["locations"]["inception_dir"])
    backend_dir = path.realpath(get_settings()["locations"]["backend_dir"])

    if not inception_init_check():
        try:
            subprocess.run(args=["terraform", "init", "-backend-config",
                                 f"{backend_dir}/pl-dev-inception.tfvars"],
                           cwd=site_dir, stdout=subprocess.DEVNULL,
                           stderr=subprocess.STDOUT)
        except subprocess.CalledProcessError:
            print(Fore.RED + "There was an error while doing the initializing")
            exit(1)

        try:
            with open(f"{site_dir}/.terraform/environment", "w") as f:
                f.write("dev")
        except FileNotFoundError:
            print(Fore.RED + "The file that manages the environments could not"
                  " be created.")
            exit(1)

        reload_direnv()
        print(Fore.GREEN + "The initialization was successfull!")

    else:
        print(Fore.YELLOW + "The initialization was already done.")

File: terrabutler/internal/inception/inception.go
package inception

import (
	"fmt"
	"os"
	"os/exec"
	"terrabutler/internal/logger"
	"terrabutler/internal/settings"
	"terrabutler/internal/utils"

	"go.uber.org/zap"
)

func inception_init_check() bool {
	dir := utils.Paths["inception"]
	if _, err := os.Stat(dir + "/.terraform"); !os.IsNotExist(err) {
		if _, err2 := os.Stat(dir + "/.terraform/environment"); !os.IsNotExist(err2) {
			return true
		}
	}
	return false
}

func Init_needed() {

	if !inception_init_check() {
		logger.Zap.Error("The initialization hasn't been made yet. Please execute the following command to initialize it: terrabutler init")
		os.Exit(1)
	}
}

func Init() {
	org := settings.Conf.String("general.organization")
	default_env_name := settings.Conf.String("environments.default.name")
	inception_dir := utils.Paths["inception"]
	backend_dir := utils.Paths["backends"]

	if !inception_init_check() {

		//run Terraform init -backend-config LOCATION
		os.Chdir(inception_dir)

		//Verifies that terraform exist in the current dir
		_, err := exec.LookPath("terraform")
		if err != nil {
			logger.Zap.Error("No Terraform executable found.")
			os.Exit(1)
		}

		command := []string{"terraform", "init", "-backend-config", backend_dir + "/" + org + "-" + default_env_name + "-inception.tfvars"}

		logger.Zap.Debug(fmt.Sprintf("Executing %s command with args: %v", command[0], command[1:]))

		//Runs the inception init command
		cmd := exec.Command(command[0], command[1:]...)
		cmd.Stderr = os.Stderr
		err = cmd.Run()
		//Show error message
		if err != nil {
			logger.Zap.Error("There was an error while doing the initialization", zap.Error(err))
			os.Exit(1)
		}
		//Try opening the new inception dir and create a file and write a file
		f, err := os.OpenFile(inception_dir+"/.terraform/environment", os.O_CREATE|os.O_WRONLY, 0644)
		if err != nil {
			logger.Zap.Error("The file that manages the environments could not be created.", zap.Error(err))
			os.Exit(1)
		}
		_, err = f.Write([]byte(default_env_name))
		if err != nil {
			logger.Zap.Error("Writing on the file that manages the environments wasn't possible.")
			os.Exit(1)
		}
		f.Close()

		//If all is ok, display successfully inception
		logger.Zap.Info("The initialization was successful!")
	} else {
		logger.Zap.Warn("The initialization was already done.")
	}

}
